src/rt/lifetime.d

```diff
@@ -281,15 +283,7 @@ bool __setArrayAllocLength(ref BlkInfo info, size_t newlength, bool isshared, co
         {
             if(isshared)
             {
-                try synchronized(typeid(ArrayAllocLengthLock))
-                {
-                    if(*length == cast(ubyte)oldlength)
-                        *length = cast(ubyte)newlength;
-                    else
-                        return false;
-                }
-                catch (Throwable t)
-                    assert(0, "Failed to synchronize.");
+                return cas(cast(shared)length, cast(ubyte)oldlength, cast(ubyte)newlength);
```

Comment:
Memory allocation has no real side effect. Before you allocate there is a GC with infinite memory, afterwards it's still a GC with infinite memory. When you lock a mutex you change it's state. Now one could argue, that many mutex implementations, e.g. a spinlock, only change the state of the mutex itself, but many mutex implementation will change some additional global variables, therefor synchronized can't be pure.

Also this change removes a severe contention bottleneck, so let's go on.